### PR TITLE
Refactored SMT realization

### PR DESCRIPTION
**Pull request by RuslanProgrammer**

### CODE REVIEW

File: circuits/Smt.circom [hunk 1]
```diff
@@ -1,167 +1,255 @@
...
+    // Check that if checking for exclusion and auxIsEmpty==0 then key!=auxKey
+    component areKeyEquals = IsEqual();
+    areKeyEquals.in[0] <== auxKey;
+    areKeyEquals.in[1] <== key;
+
+    component keysOk = MultiAND(3);
+    keysOk.in[0] <== isExclusion;
+    keysOk.in[1] <== inverse(auxIsEmpty);
+    keysOk.in[2] <== areKeyEquals.out;
+    keysOk.out === 0;
```
Comment (Arvolear) [inline circuits/Smt.circom:252]:
```suggestion
    keysOk.in[2] <== inverse(areKeyEquals.out);
    keysOk.out === 1;
```

??

File: circuits/Smt.circom [hunk 2]
```diff
@@ -1,167 +1,255 @@
...
+    signal input isDesiredDepth;
+    signal input isExclusion;
 
-    signal output st_top;
-    signal output st_inew;
+    signal input previousMiddle;
+    signal input previousEmpty;
+    signal input previousAuxLeaf;
+    signal input previousLeaf;
+    signal input previous_na; // TODO: Should we remove this?
```
Comment (Arvolear) [inline circuits/Smt.circom:92]: Let's remove these

File: circuits/Smt.circom (hunk 1 above)
Comment (RuslanProgrammer) [reply circuits/Smt.circom:252]: In this case, two other values have to be true always.

**Review by Arvolear — APPROVED** [review]
LGTM!

**Pull request merged**